### PR TITLE
Add persistedState used by table states to keep filters during unmount

### PR DESCRIPTION
**Pull request by nsams**
Problem appears when state (filter, paging, sort) is created inside a Stack page which is unmounted
when showing detail form. On remount (when user goes back) the filters are reset to default values.

This is now worked around by creating a persistent state id at stack level and passing it down to the
table. usePersistedStore backs up the value on unmount and reuses that value on remount.

Persisted might be missleading as it is not persisted across page reloads, only across mount/unmounts.

However we could probably want to persist (optionally) to localstorage to support also that (if needed)

Also add a story showing the problem, plus another story for an alternative solution (where all states are at stack level)

### CODE REVIEW

File: packages/react-admin-core/src/table/useCreatePersistedStateId.tsx [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import * as React from "react";
+import useConstant from "use-constant";
+
+const UUID = require("uuid");
+
+export function useCreatePersistedStateId(): string {
```
Comment (johnnyomair) [inline packages/react-admin-core/src/table/useCreatePersistedStateId.tsx]: I'd probably call this `usePersistedStateId`, as the ID is created only once.

File: packages/react-admin-core/src/table/usePersistedState.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    const [state, setState] = React.useState<T>(v);
+    if (stateId) delete allStates[stateId]; // delete from allStates as the component is mounted now and handles it's state itself
+
+    React.useEffect(() => {
+        return () => {
+            // on unmount we backukp the current state into allStates
+            if (stateId) allStates[stateId] = state;
+        };
+    }, [state]);
```
Comment (johnnyomair) [inline packages/react-admin-core/src/table/usePersistedState.tsx]: Do you need the "hook dependency" `state` for this hook? As the action is performed on unmount only, it should be enough when the `useEffect` hook is triggered only on mount.

File: packages/react-admin-core/src/table/usePersistedState.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    const stateId = options.persistedStateId;
+
+    const v = (stateId && allStates[stateId]) || defaultValue;
+
+    const [state, setState] = React.useState<T>(v);
+    if (stateId) delete allStates[stateId]; // delete from allStates as the component is mounted now and handles it's state itself
+
+    React.useEffect(() => {
+        return () => {
+            // on unmount we backukp the current state into allStates
```
Comment (johnnyomair) [inline packages/react-admin-core/src/table/usePersistedState.tsx]:
```suggestion
            // on unmount we backup the current state into allStates
```

File: packages/react-admin-core/src/table/usePersistedState.tsx (hunk 2 above)
Comment (nsams) [reply packages/react-admin-core/src/table/usePersistedState.tsx]: ja, weil sonst speichert er den alten (=initialen) state

**Pull request merged**